File: scripts/dbx-dogfood-loader.py
#!/usr/bin/env python3
"""dbx-dogfood-loader.py — the Databricks leg's day loader for
examples/github_activity/ (docs/outcomes/20260912-databricks-dogfood-spine/
phases/03-plan.md).

External to smelt, exactly as scripts/bq-dogfood-loader.sh is for BigQuery —
smelt orders the run, it does not author the load. It never restates
examples/github_activity/load_day.sh's redelivery rule: the modulus is
*parsed* out of that script at runtime, so a re-pin of load_day.sh cannot
silently drift out of step with what lands in Unity Catalog.

Modes:
    --emit-ddl                          Delta DDL for the two raw tables and
                                         the per-day ledger table. No network.
    --apply-ddl                         Executes the same statements
                                         --emit-ddl prints, against the live
                                         workspace (needs a credential).
    --emit-sql --date D                 Describes day D's append in
                                         catalog-qualified terms. No network.
    --emit-slice-sql --date D           The DuckDB-executable SELECT queries
                                         (event-time and arrival) that produce
                                         day D's rows straight from the
                                         fixture — the per-PR identity gate
                                         compares these to load_day.sh's own
                                         output with no workspace. No network.
    --date D                            Execute: read day D's slice from the
                                         fixture via DuckDB, hand it to
                                         Databricks Connect as Arrow, and
                                         append it into Unity Catalog. Rows
                                         cross via the Arrow load path
                                         (DatabricksAdapter.load_arrow_table)
                                         — Databricks Connect never reads a
                                         host-visible file, since serverless
                                         compute shares no filesystem with the
                                         client
                                         (docs/specs/multi_backend.md
                                         §"Loading data into a backend").
    --date D --dry-run-store <dir>      Same idempotence guard and fixture
                                         read as --date D, but records the day
                                         in a local ledger file under <dir>
                                         instead of touching a real
                                         workspace — proves the guard and the
                                         read path with no credential.
    --next-day [--dry-run-store <dir>]  Resolves the day itself: the earliest
                                         fixture day the ledger (the live
                                         `_loader_days` table, or the dry-run
                                         store) has not yet recorded, then
                                         loads it exactly as --date D would.
                                         A scheduled run has no real calendar
                                         relationship to the fixed historical
                                         fixture, so it must advance its own
                                         ledger rather than trust wall-clock
                                         time. Exits 0 with a message, never
                                         an error, once the fixture is
                                         exhausted.

The idempotence guard always runs BEFORE any frame is built (matching
load_day.sh's own `_loader_days` check): an already-loaded day is a no-op
that touches neither the fixture nor Unity Catalog / the dry-run store.

DuckDB access prefers the `duckdb` Python module and falls back to the
`duckdb` CLI binary only when the module is not importable — a serverless
Databricks Python environment installs Python packages but has no CLI binary
on PATH, so the module path is what actually runs in production; the CLI
fallback exists for a local shell that has the binary but not the package.
"""

import argparse
import json
import os
import re
import subprocess
import sys

REPO_ROOT = os.path.abspath(os.path.join(os.path.dirname(__file__), ".."))
LOAD_DAY_SH = os.path.join(REPO_ROOT, "examples/github_activity/load_day.sh")
SAMPLE_PARQUET = os.path.join(
    REPO_ROOT, "examples/github_activity/seeds/github_events_sample.parquet"
)

CATALOG = os.environ.get("SMELT_DBX_CATALOG", "workspace")
SCHEMA = os.environ.get("SMELT_DBX_SCHEMA", "smelt_dogfood")
EVENTS_TABLE = f"{CATALOG}.{SCHEMA}.github_events"
ARRIVAL_TABLE = f"{CATALOG}.{SCHEMA}.github_events_arrival"
LEDGER_TABLE = f"{CATALOG}.{SCHEMA}._loader_days"


def parsed_modulus():
    """The redelivery modulus, parsed out of load_day.sh rather than restated."""
    with open(LOAD_DAY_SH, encoding="utf-8") as f:
        text = f.read()
    m = re.search(r"MOD\(CAST\(id AS BIGINT\), (\d+)\)", text)
    if not m:
        raise RuntimeError(f"could not parse the redelivery modulus out of {LOAD_DAY_SH}")
    return int(m.group(1))


def events_select_sql(date, modulus):
    return (
        f"SELECT * FROM read_parquet('{SAMPLE_PARQUET}')\n"
        f"WHERE CAST(created_at AS DATE) = DATE '{date}'\n"
        "UNION ALL\n"
        f"SELECT * FROM read_parquet('{SAMPLE_PARQUET}')\n"
        f"WHERE CAST(created_at AS DATE) = DATE '{date}' - INTERVAL 1 DAY\n"
        f"  AND MOD(CAST(id AS BIGINT), {modulus}) = 0"
    )


def arrival_select_sql(date, modulus):
    return (
        f"SELECT *, DATE '{date}' AS ingested_date FROM read_parquet('{SAMPLE_PARQUET}')\n"
        f"WHERE CAST(created_at AS DATE) = DATE '{date}'\n"
        "UNION ALL\n"
        f"SELECT *, DATE '{date}' AS ingested_date FROM read_parquet('{SAMPLE_PARQUET}')\n"
        f"WHERE CAST(created_at AS DATE) = DATE '{date}' - INTERVAL 1 DAY\n"
        f"  AND MOD(CAST(id AS BIGINT), {modulus}) = 0"
    )


def cmd_emit_slice_sql(date):
    modulus = parsed_modulus()
    print("-- events")
    print(events_select_sql(date, modulus) + ";")
    print()
    print("-- arrival")
    print(arrival_select_sql(date, modulus) + ";")


def ddl_statements():
    """The Delta DDL for the two raw tables and the per-day ledger table, as a
    list of statements — the single source both --emit-ddl and --apply-ddl
    consume, so the two modes cannot drift apart."""
    return [
        f"CREATE TABLE IF NOT EXISTS {EVENTS_TABLE} (\n"
        f"  id STRING,\n"
        f"  type STRING,\n"
        f"  created_at TIMESTAMP,\n"
        f"  actor_id BIGINT,\n"
        f"  actor_login STRING,\n"
        f"  repo_id BIGINT,\n"
        f"  repo_name STRING,\n"
        f"  org_id BIGINT,\n"
        f"  public BOOLEAN,\n"
        f"  payload STRING\n"
        ")\n"
        "USING DELTA",
        f"CREATE TABLE IF NOT EXISTS {ARRIVAL_TABLE} (\n"
        f"  id STRING,\n"
        f"  type STRING,\n"
        f"  created_at TIMESTAMP,\n"
        f"  actor_id BIGINT,\n"
        f"  actor_login STRING,\n"
        f"  repo_id BIGINT,\n"
        f"  repo_name STRING,\n"
        f"  org_id BIGINT,\n"
        f"  public BOOLEAN,\n"
        f"  payload STRING,\n"
        f"  ingested_date DATE\n"
        ")\n"
        "USING DELTA",
        f"CREATE TABLE IF NOT EXISTS {LEDGER_TABLE} (\n"
        "  day DATE\n"
        ")\n"
        "USING DELTA",
    ]


def cmd_emit_ddl():
    print(
        f"-- Delta DDL for the Databricks dogfood leg "
        f"(docs/outcomes/20260912-databricks-dogfood-spine/phases/03-plan.md).\n"
        f"-- Catalog-qualified only — the fixture never appears here; rows\n"
        f"-- cross via the Arrow load path (see --date D to execute)."
    )
    for stmt in ddl_statements():
        print(stmt + ";\n")


def cmd_apply_ddl():
    from smelt.databricks_adapter import DatabricksAdapter

    host = os.environ.get("SMELT_DBX_HOST")
    if not host:
        print("SMELT_DBX_HOST is not set — source scripts/dbx-dogfood-env.sh first", file=sys.stderr)
        sys.exit(1)
    token = os.environ.get("SMELT_DBX_TOKEN")

    adapter = DatabricksAdapter(host, catalog=CATALOG, token=token)
    try:
        for stmt in ddl_statements():
            adapter.execute_sql_no_result(stmt)
        print("applied DDL")
    finally:
        adapter.close()


def cmd_emit_sql(date):
    modulus = parsed_modulus()
    print(
        f"-- Databricks load for day {date}: rows cross via the Arrow load path\n"
        "-- (DatabricksAdapter.load_arrow_table), never a host-visible file —\n"
        "-- serverless compute shares no filesystem with the client\n"
        '-- (docs/specs/multi_backend.md §"Loading data into a backend").\n'
        "--\n"
        f"-- Real day-{date} rows plus a deterministic 1/{modulus} slice of the\n"
        "-- previous day's rows (the loader's declared at-least-once redelivery\n"
        "-- behaviour, MOD(CAST(id AS BIGINT), "
        f"{modulus}) = 0) are appended, via\n"
        "-- createDataFrame(...).write.saveAsTable(...), to:\n"
        f"--   {EVENTS_TABLE}\n"
        f"--   {ARRIVAL_TABLE} (stamped ingested_date = DATE '{date}')\n"
        "--\n"
        f"-- Use --date {date} to execute, or --emit-slice-sql --date {date} to\n"
        "-- see the DuckDB-side SELECT that produces the rows."
    )


def _duckdb_module():
    """The importable `duckdb` Python module, or None. Cached per-process:
    called on every query, and the import itself is the expensive part when
    it fails (a full sys.path scan)."""
    if not hasattr(_duckdb_module, "_cached"):
        try:
            import duckdb as duckdb_module

            _duckdb_module._cached = duckdb_module
        except ImportError:
            _duckdb_module._cached = None
    return _duckdb_module._cached


def duckdb_rows(sql):
    """Runs `sql` against the fixture and returns a list of row tuples, via
    the `duckdb` Python module when importable, else the `duckdb` CLI. One
    SQL definition, two access paths — a serverless Databricks Python
    environment has the module but no CLI binary on PATH."""
    duckdb_module = _duckdb_module()
    if duckdb_module is not None:
        con = duckdb_module.connect(":memory:")
        try:
            return con.execute(sql).fetchall()
        finally:
            con.close()
    proc = subprocess.run(
        ["duckdb", ":memory:", "-json", "-c", sql],
        capture_output=True,
        check=False,
    )
    if proc.returncode != 0:
        raise RuntimeError(f"duckdb query failed: {proc.stderr.decode(errors='replace')}")
    rows = json.loads(proc.stdout.decode())
    return [tuple(row.values()) for row in rows]


def duckdb_query_arrow(sql):
    """Runs `sql` against the fixture and returns Arrow IPC stream bytes, via
    the `duckdb` Python module when importable, else the `duckdb` CLI.

    The CLI path needs an explicit `LOAD arrow`: `arrow` moved out of
    DuckDB's core/autoload-known extension set into the community
    repository, so unlike `parquet` it is never autoloaded even once
    `INSTALL ... FROM community` has cached it on disk
    (`duckdb :memory: -c "INSTALL arrow FROM community"`, a one-time step
    this function does not perform itself)."""
    duckdb_module = _duckdb_module()
    if duckdb_module is not None:
        import pyarrow as pa

        con = duckdb_module.connect(":memory:")
        try:
            table = con.execute(sql).arrow()
        finally:
            con.close()
        sink = pa.BufferOutputStream()
        writer = pa.ipc.new_stream(sink, table.schema)
        writer.write_table(table)
        writer.close()
        return sink.getvalue().to_pybytes()
    proc = subprocess.run(
        [
            "duckdb",
            ":memory:",
            "-c",
            f"LOAD arrow; COPY ({sql}) TO '/dev/stdout' (FORMAT arrow)",
        ],
        capture_output=True,
        check=False,
    )
    if proc.returncode != 0:
        raise RuntimeError(f"duckdb query failed: {proc.stderr.decode(errors='replace')}")
    return proc.stdout


def duckdb_scalar(sql):
    return duckdb_rows(sql)[0][0]


def fixture_days():
    """Every distinct day the fixture holds, ascending — the pool `--next-day`
    picks its candidate from."""
    rows = duckdb_rows(
        f"SELECT DISTINCT CAST(created_at AS DATE) AS d FROM read_parquet('{SAMPLE_PARQUET}') "
        "ORDER BY d"
    )
    return [str(r[0]) for r in rows]


def ledger_path(store_dir):
    return os.path.join(store_dir, "loader_days.txt")


def dry_run_already_loaded(store_dir, date):
    path = ledger_path(store_dir)
    if not os.path.exists(path):
        return False
    with open(path, encoding="utf-8") as f:
        return date in {line.strip() for line in f if line.strip()}


def dry_run_record(store_dir, date):
    with open(ledger_path(store_dir), "a", encoding="utf-8") as f:
        f.write(date + "\n")


def dry_run_loaded_days(store_dir):
    path = ledger_path(store_dir)
    if not os.path.exists(path):
        return set()
    with open(path, encoding="utf-8") as f:
        return {line.strip() for line in f if line.strip()}


def next_unloaded_day(loaded_days):
    """The earliest fixture day not in `loaded_days`, or None once the fixture
    is exhausted — the one predicate both the dry-run and live `--next-day`
    paths share."""
    for day in fixture_days():
        if day not in loaded_days:
            return day
    return None


def cmd_next_day_dry_run(store_dir):
    os.makedirs(store_dir, exist_ok=True)
    day = next_unloaded_day(dry_run_loaded_days(store_dir))
    if day is None:
        print("no fixture days remain to load", file=sys.stderr)
        return
    cmd_dry_run_store(day, store_dir)


def cmd_next_day():
    from smelt.databricks_adapter import DatabricksAdapter

    host = os.environ.get("SMELT_DBX_HOST")
    if not host:
        print("SMELT_DBX_HOST is not set — source scripts/dbx-dogfood-env.sh first", file=sys.stderr)
        sys.exit(1)
    token = os.environ.get("SMELT_DBX_TOKEN")

    adapter = DatabricksAdapter(host, catalog=CATALOG, token=token)
    try:
        loaded_days = set()
        if adapter.table_exists(LEDGER_TABLE):
            result = adapter.execute_sql(f"SELECT day FROM {LEDGER_TABLE}")
            loaded_days = {str(v.as_py()) for v in result.column("day")}
        day = next_unloaded_day(loaded_days)
    finally:
        adapter.close()

    if day is None:
        print("no fixture days remain to load", file=sys.stderr)
        return
    cmd_execute(day)


def cmd_dry_run_store(date, store_dir):
    os.makedirs(store_dir, exist_ok=True)
    # The guard runs BEFORE any frame is built — an already-loaded day never
    # touches the fixture, matching load_day.sh's own ordering.
    if dry_run_already_loaded(store_dir, date):
        print(f"day {date} already loaded, skipping", file=sys.stderr)
        return

    modulus = parsed_modulus()
    events_count = duckdb_scalar(f"SELECT count(*) AS c FROM ({events_select_sql(date, modulus)}) t")
    arrival_count = duckdb_scalar(
        f"SELECT count(*) AS c FROM ({arrival_select_sql(date, modulus)}) t"
    )
    dry_run_record(store_dir, date)
    print(f"loaded day {date} (dry run): {events_count} events, {arrival_count} arrival rows")


def cmd_execute(date):
    from smelt.databricks_adapter import DatabricksAdapter

    host = os.environ.get("SMELT_DBX_HOST")
    if not host:
        print("SMELT_DBX_HOST is not set — source scripts/dbx-dogfood-env.sh first", file=sys.stderr)
        sys.exit(1)
    token = os.environ.get("SMELT_DBX_TOKEN")

    adapter = DatabricksAdapter(host, catalog=CATALOG, token=token)
    try:
        if adapter.table_exists(LEDGER_TABLE):
            already = adapter.execute_sql(
                f"SELECT count(*) AS c FROM {LEDGER_TABLE} WHERE day = DATE '{date}'"
            )
            if already.column("c")[0].as_py() > 0:
                print(f"day {date} already loaded, skipping", file=sys.stderr)
                return

        modulus = parsed_modulus()
        events_bytes = duckdb_query_arrow(events_select_sql(date, modulus))
        arrival_bytes = duckdb_query_arrow(arrival_select_sql(date, modulus))

        adapter.load_arrow_table(events_bytes, EVENTS_TABLE, mode="append")
        adapter.load_arrow_table(arrival_bytes, ARRIVAL_TABLE, mode="append")
        adapter.execute_sql_no_result(f"INSERT INTO {LEDGER_TABLE} VALUES (DATE '{date}')")
        print(f"loaded day {date}")
    finally:
        adapter.close()


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--emit-ddl", action="store_true")
    parser.add_argument("--apply-ddl", action="store_true")
    parser.add_argument("--emit-sql", action="store_true")
    parser.add_argument("--emit-slice-sql", action="store_true")
    parser.add_argument("--next-day", action="store_true")
    parser.add_argument("--date")
    parser.add_argument("--dry-run-store")
    args = parser.parse_args()

    if args.emit_ddl:
        cmd_emit_ddl()
        return
    if args.apply_ddl:
        cmd_apply_ddl()
        return
    if args.emit_sql:
        if not args.date:
            parser.error("--emit-sql requires --date")
        cmd_emit_sql(args.date)
        return
    if args.emit_slice_sql:
        if not args.date:
            parser.error("--emit-slice-sql requires --date")
        cmd_emit_slice_sql(args.date)
        return
    if args.next_day:
        if args.date:
            parser.error("--next-day resolves its own date — do not pass --date")
        if args.dry_run_store:
            cmd_next_day_dry_run(args.dry_run_store)
        else:
            cmd_next_day()
        return
    if not args.date:
        parser.error("--date is required")
    if args.dry_run_store:
        cmd_dry_run_store(args.date, args.dry_run_store)
        return
    cmd_execute(args.date)


if __name__ == "__main__":
    main()

File: crates/smelt-cli/tests/dbx_dogfood_loader.rs
#![cfg(feature = "duckdb")]
//! Tests for `scripts/dbx-dogfood-loader.py`/`.sh`, the pinned
//! `databricks-connect` client venv, and `scripts/dbx-dogfood-env.sh`
//! (`docs/outcomes/20260912-databricks-dogfood-spine/phases/03-plan.md`).
//!
//! The loader is external to smelt — smelt's source declaration is the
//! contract it is trusted against, exactly as `scripts/bq-dogfood-loader.sh`
//! is for BigQuery (`crates/smelt-cli/tests/github_activity_loader.rs`). Its
//! `--emit-ddl`/`--emit-sql`/`--emit-slice-sql` modes touch no network and
//! need no `databricks-connect` package or workspace, so this file proves
//! the loader is byte-for-byte faithful to
//! `examples/github_activity/load_day.sh`'s own redelivery rule with no
//! cloud in the loop.

use std::path::{Path, PathBuf};
use std::process::{Command, Output};
use tempfile::TempDir;

mod github_activity_support;
use github_activity_support::{load_day, stage_workspace, FIXTURE_DAYS};

fn repo_root() -> PathBuf {
    Path::new(env!("CARGO_MANIFEST_DIR"))
        .parent()
        .unwrap()
        .parent()
        .unwrap()
        .to_path_buf()
}

fn loader_py() -> PathBuf {
    repo_root().join("scripts/dbx-dogfood-loader.py")
}

fn load_day_sh() -> PathBuf {
    repo_root().join("examples/github_activity/load_day.sh")
}

fn run_loader(args: &[&str]) -> Output {
    Command::new("python3")
        .arg(loader_py())
        .args(args)
        .current_dir(repo_root())
        .output()
        .unwrap_or_else(|e| panic!("failed to spawn dbx-dogfood-loader.py: {e}"))
}

fn duckdb_scalar(sql: &str) -> i64 {
    let out = Command::new("duckdb")
        .args([":memory:", "-json", "-c", sql])
        .output()
        .unwrap_or_else(|e| panic!("failed to spawn duckdb: {e}"));
    assert!(
        out.status.success(),
        "duckdb query failed: {}\nsql: {sql}",
        String::from_utf8_lossy(&out.stderr)
    );
    let stdout = String::from_utf8_lossy(&out.stdout);
    let rows: serde_json::Value = serde_json::from_str(stdout.trim())
        .unwrap_or_else(|e| panic!("invalid JSON: {e}\n{stdout}"));
    rows[0]
        .as_object()
        .and_then(|obj| obj.values().next())
        .and_then(|v| v.as_i64())
        .unwrap_or_else(|| panic!("expected one scalar column in query result: {rows}"))
}

fn duckdb_scalar_against(db: &Path, sql: &str) -> i64 {
    let out = Command::new("duckdb")
        .arg(db)
        .args(["-json", "-c", sql])
        .output()
        .unwrap_or_else(|e| panic!("failed to spawn duckdb: {e}"));
    assert!(
        out.status.success(),
        "duckdb query failed: {}\nsql: {sql}",
        String::from_utf8_lossy(&out.stderr)
    );
    let stdout = String::from_utf8_lossy(&out.stdout);
    let rows: serde_json::Value = serde_json::from_str(stdout.trim())
        .unwrap_or_else(|e| panic!("invalid JSON: {e}\n{stdout}"));
    rows[0]
        .as_object()
        .and_then(|obj| obj.values().next())
        .and_then(|v| v.as_i64())
        .unwrap_or_else(|| panic!("expected one scalar column in query result: {rows}"))
}

/// Splits `--emit-slice-sql`'s output into (events_query, arrival_query),
/// delimited by the `-- events` / `-- arrival` markers the loader prints.
fn split_slice_sql(emitted: &str) -> (String, String) {
    let events_marker = "-- events";
    let arrival_marker = "-- arrival";
    let events_start = emitted
        .find(events_marker)
        .unwrap_or_else(|| panic!("missing '{events_marker}' marker in:\n{emitted}"));
    let arrival_start = emitted
        .find(arrival_marker)
        .unwrap_or_else(|| panic!("missing '{arrival_marker}' marker in:\n{emitted}"));
    assert!(events_start < arrival_start);
    let events = emitted[events_start + events_marker.len()..arrival_start]
        .trim()
        .trim_end_matches(';')
        .to_string();
    let arrival = emitted[arrival_start + arrival_marker.len()..]
        .trim()
        .trim_end_matches(';')
        .to_string();
    (events, arrival)
}

fn parsed_modulus() -> u32 {
    let text = std::fs::read_to_string(load_day_sh()).expect("read load_day.sh");
    let line = text
        .lines()
        .find(|l| l.contains("MOD(CAST(id AS BIGINT),"))
        .expect("load_day.sh names its redelivery modulus");
    let after = line
        .split("MOD(CAST(id AS BIGINT),")
        .nth(1)
        .expect("modulus follows MOD(CAST(id AS BIGINT),");
    after
        .trim_start()
        .split(')')
        .next()
        .expect("modulus is followed by a closing paren")
        .trim()
        .parse()
        .expect("modulus is an integer")
}

#[test]
fn slice_matches_load_day_rows_for_every_fixture_day() {
    for day in FIXTURE_DAYS {
        let tmp = TempDir::new().expect("tempdir");
        let (_workspace, db, sample) = stage_workspace(tmp.path());
        load_day(&db, &sample, day, None);

        let out = run_loader(&["--emit-slice-sql", "--date", day]);
        assert!(
            out.status.success(),
            "day {day} stderr: {}",
            String::from_utf8_lossy(&out.stderr)
        );
        let emitted = String::from_utf8_lossy(&out.stdout).to_string();
        let (events_sql, _) = split_slice_sql(&emitted);

        let left = duckdb_scalar_against(
            &db,
            &format!(
                "SELECT count(*) AS c FROM (SELECT * FROM main.sources_raw_github_events \
                 EXCEPT ALL ({events_sql})) t"
            ),
        );
        let right = duckdb_scalar_against(
            &db,
            &format!(
                "SELECT count(*) AS c FROM (({events_sql}) \
                 EXCEPT ALL SELECT * FROM main.sources_raw_github_events) t"
            ),
        );
        assert_eq!(
            left, 0,
            "day {day}: loaded rows not covered by the emitted slice"
        );
        assert_eq!(
            right, 0,
            "day {day}: emitted slice rows not present in load_day.sh's output"
        );

        let loaded_count = duckdb_scalar_against(
            &db,
            "SELECT count(*) AS c FROM main.sources_raw_github_events",
        );
        let slice_count =
            duckdb_scalar_against(&db, &format!("SELECT count(*) AS c FROM ({events_sql}) t"));
        assert_eq!(loaded_count, slice_count, "day {day}: row counts diverge");
    }
}

#[test]
fn arrival_slice_stamps_ingested_date_like_load_day() {
    let day = "2026-08-06";
    let tmp = TempDir::new().expect("tempdir");
    let (_workspace, db, sample) = stage_workspace(tmp.path());
    load_day(&db, &sample, day, None);

    let out = run_loader(&["--emit-slice-sql", "--date", day]);
    assert!(out.status.success());
    let emitted = String::from_utf8_lossy(&out.stdout).to_string();
    let (_, arrival_sql) = split_slice_sql(&emitted);

    let left = duckdb_scalar_against(
        &db,
        &format!(
            "SELECT count(*) AS c FROM (SELECT * FROM main.sources_raw_github_events_arrival \
             EXCEPT ALL ({arrival_sql})) t"
        ),
    );
    let right = duckdb_scalar_against(
        &db,
        &format!(
            "SELECT count(*) AS c FROM (({arrival_sql}) \
             EXCEPT ALL SELECT * FROM main.sources_raw_github_events_arrival) t"
        ),
    );
    assert_eq!(left, 0);
    assert_eq!(right, 0);

    let stamped = duckdb_scalar_against(
        &db,
        &format!("SELECT count(*) AS c FROM ({arrival_sql}) t WHERE ingested_date <> DATE '{day}'"),
    );
    assert_eq!(
        stamped, 0,
        "every row in the arrival slice must be stamped ingested_date = {day}"
    );
}

#[test]
fn first_fixture_day_has_an_empty_redelivery_arm() {
    let day = FIXTURE_DAYS[0];
    let out = run_loader(&["--emit-slice-sql", "--date", day]);
    assert!(out.status.success());
    let emitted = String::from_utf8_lossy(&out.stdout).to_string();
    let (events_sql, _) = split_slice_sql(&emitted);

    let sample = repo_root().join("examples/github_activity/seeds/github_events_sample.parquet");
    let plain_day_count = duckdb_scalar(&format!(
        "SELECT count(*) AS c FROM read_parquet('{}') WHERE CAST(created_at AS DATE) = DATE '{day}'",
        sample.display()
    ));
    let slice_count = duckdb_scalar(&format!("SELECT count(*) AS c FROM ({events_sql}) t"));
    assert_eq!(
        plain_day_count, slice_count,
        "the earliest fixture day has no D-1 rows in the sample, so its redelivery arm must \
         match zero rows and the loader needs no special case"
    );
}

#[test]
fn redelivery_modulus_is_parsed_from_load_day_not_restated() {
    let expected = parsed_modulus();
    let out = run_loader(&["--emit-slice-sql", "--date", "2026-08-06"]);
    assert!(out.status.success());
    let emitted = String::from_utf8_lossy(&out.stdout).to_string();
    assert!(
        emitted.contains(&format!("MOD(CAST(id AS BIGINT), {expected})")),
        "expected the loader to use load_day.sh's own modulus ({expected}) in:\n{emitted}"
    );
}

#[test]
fn loader_is_idempotent_per_day() {
    let tmp = TempDir::new().expect("tempdir");
    let store = tmp.path().join("store");
    std::fs::create_dir_all(&store).expect("mkdir store");

    let out = run_loader(&[
        "--date",
        "2026-08-06",
        "--dry-run-store",
        store.to_str().unwrap(),
    ]);
    assert!(
        out.status.success(),
        "stderr: {}",
        String::from_utf8_lossy(&out.stderr)
    );
    let first_stdout = String::from_utf8_lossy(&out.stdout).to_lowercase();
    assert!(
        first_stdout.contains("loaded"),
        "first run should report it loaded: {first_stdout}"
    );
    let ledger = store.join("loader_days.txt");
    assert!(
        ledger.exists(),
        "first run should record the day in the store's ledger"
    );
    let ledger_after_first = std::fs::read_to_string(&ledger).expect("read ledger");
    assert!(ledger_after_first.contains("2026-08-06"));

    let out = run_loader(&[
        "--date",
        "2026-08-06",
        "--dry-run-store",
        store.to_str().unwrap(),
    ]);
    assert!(
        out.status.success(),
        "second invocation should also exit 0; stderr: {}",
        String::from_utf8_lossy(&out.stderr)
    );
    let combined = format!(
        "{}{}",
        String::from_utf8_lossy(&out.stdout).to_lowercase(),
        String::from_utf8_lossy(&out.stderr).to_lowercase()
    );
    assert!(
        combined.contains("already loaded"),
        "second run should report already loaded: {combined}"
    );
    let ledger_after_second = std::fs::read_to_string(&ledger).expect("read ledger");
    assert_eq!(
        ledger_after_first, ledger_after_second,
        "the second, already-loaded run must write nothing further to the ledger"
    );
}

#[test]
fn emitted_load_statements_reference_no_host_path() {
    for (args, _label) in [
        (vec!["--emit-sql", "--date", "2026-08-06"], "emit-sql"),
        (vec!["--emit-ddl"], "emit-ddl"),
    ] {
        let out = run_loader(&args);
        assert!(
            out.status.success(),
            "{args:?} stderr: {}",
            String::from_utf8_lossy(&out.stderr)
        );
        let emitted = String::from_utf8_lossy(&out.stdout).to_string();
        assert!(
            !emitted.contains("seeds/github_events_sample.parquet"),
            "{args:?} must not reference the fixture's host path:\n{emitted}"
        );
        assert!(
            !emitted.to_lowercase().contains(".parquet"),
            "{args:?} must not reference any parquet file path:\n{emitted}"
        );
        assert!(
            emitted.contains("smelt_dogfood.github_events"),
            "{args:?} should name the catalog-qualified target table:\n{emitted}"
        );
    }
}

#[test]
fn client_env_is_pinned_and_disjoint_from_the_spark_venv() {
    let pin_file = repo_root().join("scripts/dbx-dogfood-requirements.txt");
    let pins = std::fs::read_to_string(&pin_file).expect("read dbx-dogfood-requirements.txt");
    assert!(
        pins.lines().any(|l| {
            let l = l.trim();
            l.starts_with("databricks-connect==")
        }),
        "expected an exact databricks-connect== pin in:\n{pins}"
    );
    assert!(
        !pins.lines().any(|l| {
            let l = l.trim();
            l == "pyspark" || l.starts_with("pyspark==") || l.starts_with("pyspark>=")
        }),
        "the databricks-connect client must not co-install bare pyspark:\n{pins}"
    );

    let venv_script = repo_root().join("scripts/dbx-dogfood-venv.sh");
    let script = std::fs::read_to_string(&venv_script).expect("read dbx-dogfood-venv.sh");
    assert!(
        script.contains(".smelt-dbx-venv"),
        "expected the dbx venv script to target .smelt-dbx-venv"
    );
    assert!(
        !script.contains(".smelt-spark-venv"),
        "the dbx venv script must not target the local-Spark venv"
    );
}

#[test]
fn dbx_dogfood_env_exports_the_dbx_venv_and_no_secret() {
    let env_script = repo_root().join("scripts/dbx-dogfood-env.sh");
    assert!(env_script.exists());

    // A real credential config now lives at the default
    // $HOME/.config/databricks-smelt-dogfood (phase 4b's live provisioning), so
    // merely removing the three env vars falls back to reading it and the "no
    // credential config on disk" assertion below would see a real host. Point
    // SMELT_DBX_CONFIG_DIR at an empty tempdir to genuinely isolate this case.
    let tmp = TempDir::new().expect("tempdir");
    let out = Command::new("bash")
        .arg("-c")
        .arg(format!(
            "set -e; source {}; echo \"PYTHONPATH=$PYTHONPATH\"; echo \"GATE=${{SMELT_DBX_HOST-UNSET}}\"",
            env_script.display()
        ))
        .current_dir(repo_root())
        .env_remove("SMELT_DBX_HOST")
        .env_remove("SMELT_DBX_TOKEN")
        .env("SMELT_DBX_CONFIG_DIR", tmp.path())
        .output()
        .unwrap_or_else(|e| panic!("failed to source dbx-dogfood-env.sh: {e}"));
    assert!(
        out.status.success(),
        "stderr: {}",
        String::from_utf8_lossy(&out.stderr)
    );
    let stdout = String::from_utf8_lossy(&out.stdout).to_string();
    assert!(
        stdout.contains(".smelt-dbx-venv"),
        "expected the dbx venv's site-packages on PYTHONPATH:\n{stdout}"
    );
    assert!(
        !stdout.contains(".smelt-spark-venv"),
        "must never put the local-Spark venv on PYTHONPATH:\n{stdout}"
    );
    assert!(
        stdout.contains(&format!("{}", repo_root().join("python").display())),
        "expected the repo python/ package on PYTHONPATH:\n{stdout}"
    );
    assert!(
        stdout.contains("GATE=UNSET"),
        "with no credential config on disk, the live gate variable must stay unset:\n{stdout}"
    );

    let combined = format!("{}{}", stdout, String::from_utf8_lossy(&out.stderr));
    assert!(
        !combined.to_lowercase().contains("dapi"),
        "must never print a token-shaped value:\n{combined}"
    );
}
